File: 1631-path-with-minimum-effort/1631-path-with-minimum-effort.cpp
class Solution {
public:
    int minimumEffortPath(vector<vector<int>>& heights) {
        int n=heights.size();
        int m=heights[0].size();

        priority_queue<pair<int,pair<int,int>>,vector<pair<int,pair<int,int>>>,
        greater<pair<int,pair<int,int>>>>q;
        vector<vector<int>>dis(n,vector<int>(m,1e9));
        q.push({0,{0,0}});
        // int diff=0;
        dis[0][0]=0;
        int row[4]={-1,0,1,0};
        int col[4]={0,1,0,-1};

        while(!q.empty()){
            auto top=q.top();
            q.pop();
            int diff=top.first;
            int r=top.second.first;
            int c=top.second.second;
            if(r==n-1 && c==m-1){
                    return diff;
                }
            for(int i=0;i<4;i++){
                int nr=r+row[i];
                int nc=c+col[i];
                
                if(nr>=0 && nr<n && nc>=0 && nc<m){
                    int newdiff=max(abs(heights[r][c]-heights[nr][nc]),diff);
                    if(newdiff<dis[nr][nc]){
                        dis[nr][nc]=newdiff;
                        q.push({newdiff,{nr,nc}});
                    }
                }
            }
        }
        return 0;
    }
};

File: 1631-path-with-minimum-effort/1631-path-with-minimum-effort.java
class Solution {
    public int minimumEffortPath(int[][] heights) {
        int n=heights.length;
        int m=heights[0].length;

        int[][] dist=new int[n][m];
        for (int[] row : dist) Arrays.fill(row, Integer.MAX_VALUE);
        PriorityQueue<int[]> q=new PriorityQueue<>((a,b)->a[0]-b[0]);

        q.offer(new int[]{0,0,0});
        dist[0][0]=0;

        int[][] dirs={{-1,0},{0,1},{1,0},{0,-1}};

        while(!q.isEmpty()){
            int[] curr=q.poll();
            int effort=curr[0];
            int r=curr[1];
            int c=curr[2];

            if(r==n-1 && c==m-1){
                return effort;
            }

            for(int[] dir:dirs){
                // int effort=curr[0];
                int nr=r+dir[0];
                int nc=c+dir[1];


                

                if(nr>=0 && nc>=0 && nr<n && nc<m){
                    int neweffort=Math.max(effort,Math.abs(heights[r][c]-heights[nr][nc]));

                    if(neweffort<dist[nr][nc]){
                        dist[nr][nc]=neweffort;
                        q.offer(new int[]{neweffort,nr,nc});
                    }
                }
            }
        }        
        return 0;
    }
}